Run ssl_test_new in fips

mattcaswell opened the pull request:
We run ssl_test_new three times:
- once with a default library context and no providers explicitly loaded
- once with a non-default library context with the default provider loaded into it, and
- once with a non-default library context with the FIPS provider loaded into it.

In both of the latter two cases we load the "null" provider into the default context to make sure we don't accidentally pick up algorithms from there.

We do the run with a default library context because there are some tests which only work in that configuration at the moment. Namely any which involve an RSA PSS cert/key - because we do not have provider support for that at the moment. Eventually I expect we can dispense with that run and just do the second two runs.

These tests will fail since they require all the key gen PRs to be merged first as well as #11494 and #11507 (and I have not included them here). However, aside from the dependencies this should be fairly complete and can be reviewed.

slontis commented on test/ssl-tests/05-sni.cnf.in:
```diff
@@ -166,4 +166,4 @@ our @tests_tls_1_1 = (
 );
 
-push @tests, @tests_tls_1_1 unless disabled("tls1_1");
+push @tests, @tests_tls_1_1 unless disabled("tls1_1") || FIPS_MODE;
```
Will this be re-enabled at some point for FIPS_MODE? If so maybe TODO(3.0) it?

slontis commented on test/ssl-tests/20-cert-select.cnf.in:
```diff
@@ -41,13 +52,21 @@ my $server_pss_restrict_only = {
-};
+if (NO_DEFLT_LIBCTX) {
+    $server_rsa_all = {
+        "Certificate" => test_pem("servercert.pem"),
```
TODO(3.0)?

slontis commented on test/ssl-tests/20-cert-select.cnf:
```diff
@@ -5,37 +5,37 @@ num_tests = 56
 test-0 = 0-ECDSA CipherString Selection
```
This reordering sure makes this a lot harder to figure out what changed in this file..

levitte commented:
You might want to edit the subject, as #11328, #11303, #11332, and #11494 are now merged

mattcaswell commented:
> You might want to edit the subject, as #11328, #11303, #11332, and #11494 are now merged

Done

levitte commented on test/generate_ssl_tests.pl:
```diff
@@ -127,17 +127,28 @@ sub print_templates {
     local $/ = undef;
     my $content = <INPUT>;
+    $content =~ s/FIPS_MODE/$fips_mode/g;
+    $content =~ s/NO_DEFLT_LIBCTX/$no_deflt_libctx/g;
```
Ewwww.....

May I suggest this instead, then just use `$fips_mode` and `$no_deflt_libctx` directly in the `.cnf.in` files?

``` diff
diff --git a/test/generate_ssl_tests.pl b/test/generate_ssl_tests.pl
index 8cfc451fbb..ce8c22f59c 100644
--- a/test/generate_ssl_tests.pl
+++ b/test/generate_ssl_tests.pl
@@ -127,6 +127,9 @@ sub print_templates {
 # Shamelessly copied from Configure.
 sub read_config {
     my $fname = shift;
+    local $ssltests::fips_mode = $provider eq "fips";
+    local $ssltests::no_deflt_libctx =
+        $provider eq "default" || $provider eq "fips";
     open(INPUT, "< $fname") or die "Can't open input file '$fname'!\n";
     local $/ = undef;
     my $content = <INPUT>;
diff --git a/test/ssl-tests/ssltests_base.pm b/test/ssl-tests/ssltests_base.pm
index c6004fa7b2..98c9a839ad 100644
--- a/test/ssl-tests/ssltests_base.pm
+++ b/test/ssl-tests/ssltests_base.pm
@@ -17,6 +17,9 @@ sub test_pem
     return "\${ENV::TEST_CERTS_DIR}" . $dir_sep . $file,
 }
 
+our $fips_mode = 0;
+our $no_deflt_libctx = 0;
+
 our %base_server = (
     "Certificate" => test_pem("servercert.pem"),
     "PrivateKey"  => test_pem("serverkey.pem"),
```

mattcaswell replied:
Thanks - I've included that patch.

levitte commented on test/ssl-tests/02-protocol-version.cnf.in:
```diff
@@ -16,4 +16,9 @@ use warnings;
+    @tests = generate_version_tests("TLS", 1);
+} else {
+    @tests = generate_version_tests("TLS", 0);
+}
```
Why not this?

``` perl
our @tests = generate_version_tests("TLS", FIPS_MODE);
```

... or well, with the change request i just made:

``` perl
our @tests = generate_version_tests("TLS", $fips_mode);
```

mattcaswell replied:
Done

mattcaswell commented:
Fixups pushed. Please take another look.

levitte commented:
Looks much more sane from a perl perspective

paulidale commented on test/ssl-tests/14-curves.cnf.in:
```diff
@@ -10,14 +10,26 @@ use warnings;
+
+my @curves;
+
+if ($fips_mode) {
```
Would it make sense to start with the fips list and add the non-fips if `!$fips_mode`?

This avoids future updates that miss one of the if branches.

mattcaswell replied:
Yes that makes sense. Fix coming up...

slontis commented:
Woohoo..  No longer pending on #11371

mattcaswell commented:
Rebased now that #11371 has been merged. I have had to include the same libssl fixup commit that I've added to #11508 here, in order to get the tests to pass.

*Please don't review the libssl changes in this PR. Please provide any review comments on that aspect in #11508 instead*.

I've taken this out of WIP, although it cannot now be pushed until #11508 goes in.

mattcaswell commented:
Fixup push addressing @paulidale's comment above.

paulidale commented:
Travis looks unhappy :(

slontis commented:
This also needs the no_fips checks added to the .t file.
There seems to be some other issues with these tests.

mattcaswell commented:
Fixups pushed hopefully addressing the Travis failures.

I've applied a similar fix to the one that @slontis applied to #11135 

There's also a fix to ssl3_cbc_digest_record in libssl to make sure we use fetched digests where applicable. So now the fixes related to ssl3_cbc_digest_record in libssl should be reviewed - but the other changes in libssl remain out of scope of this PR (review them in #11508)

mattcaswell commented:
Darn - travis was still not happy. Another fixup pushed.

mattcaswell commented:
Sigh. Third time lucky...another fixup.

mattcaswell commented:
Travis is green!!

mattcaswell commented:
Rebased now that #11508 has gone in. No other changes were made.

levitte commented on test/recipes/80-test_ssl_new.t:
```diff
@@ -13,11 +13,21 @@ use warnings;
+
+use lib srctop_dir('Configurations');
+use lib bldtop_dir('.');
+use platform;
```
Unecessary by now...

levitte commented on test/recipes/80-test_ssl_new.t:
```diff
@@ -106,28 +117,46 @@ my %skip = (
+                '-macopt', 'digest:SHA256', '-macopt', 'hexkey:00',
+                '-section_name', 'fips_sect'])),
+       "fipsinstall");
+}
```
... and simplify this:

``` perl
unless ($no_fips) {
    ok(run(perltest(['fipsinstall.pl', bldtop_dir()])),
       'fipsinstall');
}
```

mattcaswell commented:
@levitte - I would prefer to push this as is, and modify the fipsinstall stuff with a follow on PR since this PR is now otherwise ready-to-merge.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

levitte commented:
> with a follow on PR

Okie

mattcaswell commented:
Pushed. Thanks!

mattcaswell commented:
>  I would prefer to push this as is, and modify the fipsinstall stuff with a follow on PR since this PR is now otherwise ready-to-merge.

#11580
